include/picongpu/particles/boundary/Reflecting.hpp

```diff
@@ -0,0 +1,104 @@
+/* Copyright 2021 Sergei Bastrakov, Lennert Sprenger
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+
+#include "picongpu/particles/MoveParticle.hpp"
+#include "picongpu/particles/boundary/ApplyImpl.hpp"
+#include "picongpu/particles/boundary/Kind.hpp"
+#include "picongpu/particles/boundary/Parameters.hpp"
+#include "picongpu/particles/boundary/Utility.hpp"
+#include "picongpu/particles/functor/misc/Parametrized.hpp"
+#include "picongpu/particles/manipulators/unary/FreeTotalCellOffset.hpp"
+
+#include <cstdint>
+
+
+namespace picongpu
+{
+    namespace particles
+    {
+        namespace boundary
+        {
+            struct ReflectParticleIfOutside : public functor::misc::Parametrized<Parameters>
+            {
+                //! some name is required
+                static constexpr char const* name = "reflectParticleIfOutside";
+
+                /** Process the current particle located in the given cell
+                 *
+                 * @param offsetToTotalOrigin offset of particle cell in the total domain
+                 * @param particle handle of particle to process (can be used to change attribute values)
+                 */
+                template<typename T_Particle>
+                HDINLINE void operator()(DataSpace<simDim> const& offsetToTotalOrigin, T_Particle& particle)
+                {
+                    for(uint32_t d = 0; d < simDim; d++)
+                        if((offsetToTotalOrigin[d] < m_parameters.beginInternalCellsTotal[d])
+                           || (offsetToTotalOrigin[d] >= m_parameters.endInternalCellsTotal[d]))
+                        {
+                            auto pos = particle[position_];
+
+                            if(offsetToTotalOrigin[d] >= m_parameters.endInternalCellsTotal[d])
+                                pos[d] = -pos[d];
+                            else
+                                pos[d] = 2.0_X - pos[d];
```

Comment:
Is this relative to the total domain or to the position of the particle boundary?
Did you test the new implementation for a particle boundary somewhere in the volume?

Reply:
question was resolved offline with @steindev